xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -189,7 +189,7 @@ public void handleStreamClosed(Status error) {
         resourceSubscribers.values()) {
       for (ResourceSubscriber<? extends ResourceUpdate> subscriber : subscriberMap.values()) {
         if (!subscriber.hasResult()) {
-          subscriber.onError(error);
+          subscriber.onError(error, false);
```

Comment:
Why did you special case this one? `request()` is safe to call even after the stream is dead.

I suspect this is a symptom of a larger problem: calling request on the wrong stream, because the stream changed between when the callbacks were issues and when they completed.

I think we can simplify a few things together. If AdsStreamV3 managed the counter, then it would be guaranteed to only apply to that one stream. It could then be passed as an argument to the various callbacks. Looks like it wouldn't cause problem for the tests.

```java
final class ProcessingTracker {
  private final AtomicInteger counter = new AtomicInteger(1);
  public ProcessingTracker(Runnable runnable, Executor executor) {...}
  public void startTask() {counter.incrementAndGet();}
  public void taskComplete() {
    if (counter.decrementAndGet()) {
      executor.execute(cleanup);
    }
}

final void handleRpcResponse(XdsResourceType<?> type, String versionInfo, List<Any> resources, String nonce) {
      ...
      ProcessingTracker tracker = new ProcessingTracker(() -> requestWriter.request(1), syncContext);
      xdsResponseHandler.handleResourceResponse(type, serverInfo, versionInfo, resources, nonce, tracker);
      tracker.taskComplete();
}
```

Reply:
yea because `onError` does not know it was within a handling response or on stream closed. It would mess up with the counter if it not specially handled. But I guess you solved the issue.

Reply:
hm, actually `onError` still has the same problem if we pass a tracker from `adsStream` to `xdsClient. handleResourceResponse`. Using the tracker, the stream closed will pass a null tracker. Does the null tracker bother you?